vktrace/vktrace_replay/vkreplay_vkreplay.cpp

```diff
@@ -2399,37 +2422,59 @@ VkResult vkReplay::manually_replay_vkAllocateMemory(packet_vkAllocateMemory *pPa
 
-        if (packetHeader2.packet_id == VKTRACE_TPI_VK_vkGetImageMemoryRequirements)
+        if (packetHeader1.packet_id == VKTRACE_TPI_VK_vkBindImageMemory)
             memRequirements = replayGetImageMemoryRequirements[bimPacket.image];
         else
             memRequirements = replayGetBufferMemoryRequirements[(VkBuffer)bimPacket.image];
 
+        VkDeviceSize replayAllocationSize = memRequirements.size;
+        if (bimPacket.memoryOffset > 0) {
+            // Do alignment for allocationSize in traced vkBIM/vkBBM
+            VkDeviceSize traceAllocationSize = *((VkDeviceSize *)&pPacket->pAllocateInfo->allocationSize);
+            VkDeviceSize alignedAllocationSize =
+                ((traceAllocationSize + memRequirements.alignment - 1) / memRequirements.alignment) * memRequirements.alignment;
+
+            // Do alignment for memory offset
+            replayAllocationSize +=
+                ((bimPacket.memoryOffset + memRequirements.alignment - 1) / memRequirements.alignment) * memRequirements.alignment;
+            if (alignedAllocationSize != replayAllocationSize) {
+                vktrace_LogWarning("alignedAllocationSize: 0x%x does not match replayAllocationSize: 0x%x", alignedAllocationSize,
+                                   replayAllocationSize);
+            }
+        }
+
+        doAllocate = false;
         if (!m_objMapper.m_adjustForGPU) {
             if (getMemoryTypeIdx(pPacket->device, remappedDevice, pPacket->pAllocateInfo->memoryTypeIndex, &memRequirements,
                                  &replayMemTypeIndex)) {
                 *((uint32_t *)&pPacket->pAllocateInfo->memoryTypeIndex) = replayMemTypeIndex;
-                if (*((VkDeviceSize *)&pPacket->pAllocateInfo->allocationSize) < memRequirements.size)
-                    *((VkDeviceSize *)&pPacket->pAllocateInfo->allocationSize) = memRequirements.size;
-                replayResult =
-                    m_vkFuncs.real_vkAllocateMemory(remappedDevice, pPacket->pAllocateInfo, NULL, &local_mem.replayDeviceMemory);
+                if (*((VkDeviceSize *)&pPacket->pAllocateInfo->allocationSize) < replayAllocationSize)
+                    *((VkDeviceSize *)&pPacket->pAllocateInfo->allocationSize) = replayAllocationSize;
+                doAllocate = true;
+                goto wrapItUp;
             } else {
                 vktrace_LogError("vkAllocateMemory() failed, couldn't find memory type for memoryTypeIndex");
                 return VK_ERROR_VALIDATION_FAILED_EXT;
             }
         }
+    }
 
-    } else {
-        // Platform matched exactly or there isn't a valid portablity table in the trace file,
-        // so use memoryTypeIndex from trace file.
+wrapItUp:
+
+    if (doAllocate) {
         replayResult = m_vkFuncs.real_vkAllocateMemory(remappedDevice, pPacket->pAllocateInfo, NULL, &local_mem.replayDeviceMemory);
     }
 
-wrapItUp:
+    if (saveFilePos) {
+        vktrace_FileLike_SetCurrentPosition(traceFile, saveFilePos);
+    }
```

Comment:
Current file position is not reset to saved position if return at line 2457 is executed.

Reply:
Will reset file position before the return at line 2457.  Maybe the file position should also be reset before the return at line 2374, 2396, and 2409?  Though the replay may (should?) fail when there's an error happens in vkAM.

Reply:
Agree that the replay will probably fail if errors occur, but would be best to be consistent - makes code easier to understand.

Reply:
Updated to reset file position to saved position before every return (at line 2352, 2375, 2389, 2438 in the latest code) to make it consistent.